src/Workspaces/Core/Portable/Classification/SyntaxClassification/SyntacticChangeRangeComputer.cs

```diff
@@ -0,0 +1,220 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Text;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Classification
+{
+    /// <summary>
+    /// Computes a syntactic text change range that determines the range of a document that was changed by an edit. The
+    /// portions outside this change range are guaranteed to be syntactically identical (see <see
+    /// cref="SyntaxNode.IsIncrementallyIdenticalTo"/>).  This algorithm is intended to be <em>fast</em>.  It is
+    /// technically linear in the number of nodes and tokens that may need to examined.  However, in practice, it should
+    /// operate in sub-linear time as it will bail the moment tokens don't match, and it's able to skip over matching
+    /// nodes fully without examining the contents of those nodes.  This is intended for consumers that want a
+    /// reasonably accurate change range computer, but do not want to spend an inordinate amount of time getting the
+    /// most accurate and minimal result possible.
+    /// </summary>
+    /// <remarks>
+    /// This computation is not guaranteed to be minimal.  It may return a range that includes parts that are unchanged.
```

Comment:
> This computation is not guaranteed to be minimal.

This is an exceptionally valuable algorithm relaxation which don't believe we've leveraged to the degree possible. In the event of cancellation from timeout, it would not be difficult to return the best _available_ answer (especially if you start from the left). I would actually prefer to leave this optimization for a follow-up PR, but it would be good to get an issue filed.

Reply:
In fact, by passing a TimeSpan directly into the diffing algorithm and using GetRootSynchronously, we can get rid of all the threads and use PooledStopwatch​​​​​​​ to just work until finished or the time elapses and return whatever we have. It would eliminate thread pool latency as a possible failure point.

Reply:
cool idea :)

Reply:
agreed.  i ahve made it do that and have doc'ed as such.